packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * WordPress dependencies
+ */
+import { useRegistry } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import {
+	useReducer,
+	useCallback,
+	useRef,
+	useLayoutEffect,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useCreateNavigationMenu from './use-create-navigation-menu';
+import menuItemsToBlocks from '../menu-items-to-blocks';
+
+function reducer( state, action ) {
+	switch ( action.type ) {
+		case 'RESOLVED':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: action.navMenu,
+			};
+		case 'ERROR':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: null,
+			};
+		case 'LOADING':
+			return {
+				...state,
+				isFetching: true,
+			};
+		default:
+			throw new Error( `Unexpected action type ${ action.type }` );
+	}
+}
+
+function useSafeDispatch( dispatch ) {
+	const mounted = useRef( false );
+	useLayoutEffect( () => {
+		mounted.current = true;
+		return () => ( mounted.current = false );
+	}, [] );
+	return useCallback(
+		( ...args ) => ( mounted.current ? dispatch( ...args ) : void 0 ),
+		[ dispatch ]
+	);
+}
+
+function useConvertClassicToBlockMenu( clientId ) {
+	const createNavigationMenu = useCreateNavigationMenu( clientId );
+	const registry = useRegistry();
+
+	const [ state, dispatch ] = useReducer( reducer, {
+		navMenu: null,
+		isFetching: false,
+	} );
+
+	const safeDispatch = useSafeDispatch( dispatch );
+
+	async function convertClassicMenuToBlockMenu( menuId, menuName ) {
+		const menuItemsParameters = {
+			menus: menuId,
+			per_page: -1,
+			context: 'view',
+		};
+
+		// 1. Fetch the classic Menu items.
+		const classicMenuItems = await registry
+			.resolveSelect( coreStore )
+			.getMenuItems( menuItemsParameters );
+
+		// 2. Convert the classic items into blocks.
+		const { innerBlocks } = menuItemsToBlocks( classicMenuItems );
+
+		// 3. Create the `wp_navigation` Post with the blocks.
+		const navigationMenu = await createNavigationMenu(
+			menuName,
+			innerBlocks
+		);
+
+		return navigationMenu;
+	}
+
+	const convert = useCallback(
+		( menuId, menuName ) => {
+			if ( ! menuId || ! menuName ) {
```

Comment:
You can have an async function inside useCallback, only you have to declare it separately, e.g.:

```js
const convert = useCallback(
	( menuId, menuName ) => {
		async function doTheWork() {};
		doTheWork();
	}
);
```

Reply:
Yeh but in this case the aim isn't to return `Promise` from `conver()`. We want to run a routine and then have that routine update the `state` which we can then use outside of the hook to decide on various things such as loading states.

This relieves the hook consumer from having to reimplementing the `isFetching`, `isError`...etc inside their component. Rather all this is encapsulated and we simply expose a state which the consumer can use to decide on side effects within a `useEffect`.

Reply:
All I'm saying is there are two code styles available to choose from. Note that my snippet above doesn't return a promise either.

Reply:
Oh I see. You're just saying we could move the async function outside of the `useCallback` and that would be ok.

Unless it's a blocker for you, I'd rather not do that now because I'm planning a follow up that might change things a little. I ping you when that's up.